add PodReplacementPolicy for Deployments

atiratree opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind feature

#### What this PR does / why we need it

A new deployment field, .spec.podReplacementPolicy, can be used to determine whether to replace terminating or terminated pods in deployments with new pods.

Consider the TerminationStarted and TerminationComplete policies when
rolling out a new deployment with a Recreate or RollingUpdate strategy.

Also consider these policies when scaling any deployment, either during or
after a rollout has been completed. Special care has to be taken when scaling
up multiple replicasets proportionally. If a replicaset cannot be fully scaled,
"deployment.kubernetes.io/replicaset-replicas-before-scale" annotation is added
to the replicaset so that it can be scaled again in the future once the extra
pods terminate.
  


#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

tracking issue: https://github.com/kubernetes/enhancements/issues/5882

#### Special notes for your reviewer:

- Prerequisite PR introducing  `.status.terminatingReplicas` https://github.com/kubernetes/kubernetes/pull/128546
- PR for improving deployment integration tests https://github.com/kubernetes/kubernetes/pull/134959
- Additional E2E tests also test this field in https://github.com/kubernetes/kubernetes/pull/126176


#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
A new `spec.PodReplacementPolicy` field is introduced in Deployments. When the DeploymentPodReplacementPolicy feature-gate is enabled, this field can be used to specify a policy for when to create replacement Pods during Deployment rollouts and scaling. The `"TerminationComplete`" policy recreates pods only after the old ones have fully terminated. The number of pods of the Deployment is always limited by the constraints of the Deployment strategy. This slows down the speed of rollouts, but prevents a spike in resource consumption. On the other hand, the `"TerminationStarted"` policy creates new pods before the old pods have had a chance to terminate. This can increase the speed of the rollout, but has a higher resource consumption. A new status field `.status.terminatingReplicas` is also added to Deployments and ReplicaSets to allow tracking of terminating pods.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
[KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/3973-consider-terminating-pods-deployment/README.md
```

atiratree committed:
api: add PodReplacementPolicy to Deployment

- add podReplacementPolicy field to Deployment's spec
  versioned API

atiratree committed:
add validation, field disablement and tests

atiratree committed:
update generated files and feature gates

atiratree committed:
ignore replicaset-replicas-before-scale annotation in deployment rollbacks

atiratree commented:
I have posted the implementation, but I am still missing the tests, so we will not be able to merge this feature in 1.30 due to not enough time. 

Even though it might be possible to finish this if we had enough review capacity, I would prefer not to for the following reasons:
- to allow proper review process to make sure we have not looked over something
- I have identified that we will need a new annotation `deployment.kubernetes.io/replicaset-replicas-before-scale` (please see https://github.com/kubernetes/kubernetes/pull/123670) to support partial scalling of older revision replica sets for deployments with TerminationComplete podReplacementPolicy. It would be great if we could merge this now, to make sure N-1 kubectl works with cluster that turn on the DeploymentPodReplacementPolicy feature.

atiratree commented:
rebased
/remove-lifecycle rotten

atiratree committed:
add integration tests for PodReplacementPolicy

soltysh reviewed (approved):
One question, but controller changes
/lgtm 
/approve

atiratree commented on pkg/controller/deployment/sync.go:
```diff
@@ -461,6 +490,14 @@ func (dc *DeploymentController) cleanupDeployment(ctx context.Context, oldRSs []
 		if rs.Status.Replicas != 0 || *(rs.Spec.Replicas) != 0 || rs.Generation > rs.Status.ObservedGeneration || rs.DeletionTimestamp != nil {
 			continue
 		}
+		if utilfeature.DefaultFeatureGate.Enabled(features.DeploymentPodReplacementPolicy) {
```
I think the risk is low as the default `revisionHistoryLimit`  is 10 and usually there are only 2 active replica sets during a rollout. If somebody uses lower limit or has more active replica sets they still have to  wait until `replicas == 0` so they would only be delayed by `gracefulTerminationSeconds`.

atiratree commented:
/label api-review

atiratree commented:
/test pull-kubernetes-e2e-kind-ipv6

atiratree commented:
rebased

soltysh commented:
/lgtm

atiratree commented:
rebased

mbianchidev commented:
@atiratree the assignee for the review @soltysh is marked as not available on Slack, maybe we can push for more reviewers to look into this before the code freeze deadline?

atiratree commented:
@mbianchidev the sig-apps parts have been approved and there has been no changes there from @soltysh's last review. The only thing missing from the PR is an api-review.

fedebongio commented:
/assign @kubernetes/api-reviewers 
/remove-sig api-machinery

liggitt commented on pkg/controller/deployment/util/deployment_util.go:
```diff
@@ -56,6 +58,11 @@ const (
+	// The presence of this annotation indicates that a replica set has not yet fully scaled.
+	// The value is used by the underlying replica sets to estimate their scale proportions if the deployment has surge replicas
+	// (e.g. terminating pods) and the replica set is partially scaled due to the deployment's .spec.podReplacementPolicy.
+	ReplicaSetReplicasBeforeScaleAnnotation = "deployment.kubernetes.io/replicaset-replicas-before-scale"
```
It's not clear to me why we need a new annotation on replicaset to decide how to distribute pods from a scale up across existing replicasets, especially with the skew issues that introduces to kubectl and the reconciliation issues for the annotation that introduces.

When the deployment in question has opted into TerminationComplete replacement policy, doesn't computing scaleBase as max(rs.Spec.Replicas, rs.Status.TerminatingReplicas) let us avoid overscaling without needing a new annotation?

liggitt replied:
Also, even if we do need an annotation, shouldn't all the logic around setting / using / reconciling / clearing ReplicaSetReplicasBeforeScaleAnnotation only kick in when the deployment opted into TerminationComplete? The logic is really complicated, avoiding it unless absolutely required seems better.

atiratree replied:
Deployments support proportional scaling and we need this annotation to preserve that behavior. The changes to the proportional scaling logic are outlined in the KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/3973-consider-terminating-pods-deployment/README.md#deployment-scaling-changes-and-a-new-annotation-for-replicasets

Since the scaling can be gradual, we have no way of inferring what `replicasBeforeScale` were before for all replicasets.

Using `rs.Status.TerminatingReplicas` does not work well in the scale computation:
- `TerminatingReplicas` are transient and can be removed by the replicaset controller anytime.
- We actually do not know how many terminating replicas each replicaset can have at each single point in time. The lower bound is 0, but there is no upper bound.

We have also looked at the kubectl skew impact and it is minimal: https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/3973-consider-terminating-pods-deployment/README.md#kubectl-skew

atiratree replied:
> Also, even if we do need an annotation, shouldn't all the logic around setting / using / reconciling / clearing ReplicaSetReplicasBeforeScaleAnnotation only kick in when the deployment opted into TerminationComplete? 

Similar to NewRSNewReplicas, this is done in the code above via the partialScaling variable. Should I duplicate the check as well in the `ComputeReplicaSetScaleAnnotations` function?

Clearing the annotation should be done when this feature is no longer used IMO.

liggitt replied:
The code paths are difficult enough that I *really* don't think we should change them for all pods. Limiting new logic or new sync passes to pods which have opted into the new reconcile behavior would build confidence that a bug here would not impact existing pods / replicasets / deployments on upgrade.

(this is trying to learn lessons from issues with the sidecar container feature that impacted all pods, even ones that were not using the sidecards feature)

atiratree replied:
It should be passed from above, but I agree it is safer to guard the logic in place and check if the feature is opted in. I will fix this. The clearing of the annotation should be unconditional, I think, to clean up after disabling or opting out of the feature.

atiratree commented on pkg/apis/apps/types.go:
```diff
@@ -415,6 +415,28 @@ type DeploymentSpec struct {
+	// - Recreate strategy uses TerminationComplete behavior when recreating the deployment,
+	//   but uses TerminationStarted when scaling the deployment.
+	// - RollingUpdate strategy uses TerminationStarted behavior for both rolling out and
+	//   scaling the deployments.
```
>is the scaling aspect only relevant when scaling happens in the middle of a rollout?

No it also checks for terminating pods when scaling (scale up) during a normal operation.

For example if you scale down 5 -> 2 and then scale up again 2 -> 5, it will wait for the pods to terminate (which can result in a substantial delay).  If pods terminate also for other reasons, it will wait with scaling up, until there is a capacity to run those pods.

> What does a replicaset in steady state do when one of its pods starts terminating? doesn't it replace immediately? If so, this field does not actually guarantee that the number of pods for the deployment will not exceed spec.replicas if maxSurge is 0, right?

Yes, this is only for situations where Deployment is managed, e.g. during rollouts or scaling. If there is an external disruption, it can still exceed the number of desired pods.

I mention this limitation in the docs PR: https://github.com/kubernetes/website/pull/45266. I suppose I should mention something similar here as well, right?

liggitt replied:
> Yes, this is only for situations where the Deployment is managed, e.g. during rollouts or scaling. If there is an external disruption, it may still exceed the number of desired pods.

Isn't that just as much of a problem for triggering autoscaling in undesirable ways? Building up from primitives, if replicasets could be told not to exceed their number of replicas for non-terminated pods when replacing pods, how would that affect the deployment portion of this enhancement, which builds on top of capabilities offered by replicasets?

atiratree replied:
Unfortunately, implementing this feature for ReplicaSets might have undesirable/unexpected consequences for the user. One of these is the minimum availability guaranteed by the Deployment (`.spec.replicas - .spec.strategy.rollingUpdate.maxUnavailable`. Even though we document `maxUnavailable` as a field for updates, I think people expect this to hold true throughout the whole lifecycle of the deployment. I've heard similar misunderstandings before. 

So for example, if you have a ReplicaSet, say, with 1-3 pods running on the same node, and something disrupts/evicts all of the pods. It could actually take a long time for even a single replica to become available again, resulting in application downtime.


I have included a general rationale in the KEP as to why I think it is not a good idea to implement this for RSs: https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/3973-consider-terminating-pods-deployment/README.md#consideration-for-other-controllers

Regarding the balancing, the RS can have extra pods during the down scale of the RS, so the deployment controller has to make the number of pods compliant across all RSs. So the invariant can still be broken for a single RS. A RS is a simple primitive and has no concept of a rollout.

I think the main issue bothering people today are the extra pods in rollouts, which this PR would help with. Unfortunately, I do not see an elegant solution for the external disruptions.

atiratree commented on pkg/apis/apps/types.go:
```diff
@@ -503,25 +525,43 @@ type RollingUpdateDeployment struct {
 }
 
+// DeploymentPodReplacementPolicy specifies the policy for creating Deployment Pod replacements.
+// Default is a mixed behavior depending on the DeploymentStrategy
```
https://github.com/kubernetes/kubernetes/pull/123430#discussion_r1703923704

atiratree commented on pkg/apis/apps/types.go:
```diff
@@ -503,25 +525,43 @@ type RollingUpdateDeployment struct {
 	ObservedGeneration int64
 
-	// Total number of non-terminated pods targeted by this deployment (their labels match the selector).
+	// Total number of non-terminating pods targeted by this deployment (their labels match the selector).
```
Yes, just a doc change. There is no change in the logic.

We also found out in the KEP that the comments differ in the `/api/` and `/apis/` types and we decided to unify them in this PR: https://github.com/kubernetes/enhancements/pull/4357#discussion_r1476717243

atiratree commented on pkg/apis/apps/types.go:
```diff
@@ -531,6 +571,12 @@ type DeploymentStatus struct {
 	// +optional
 	UnavailableReplicas int32
 
+	// Total number of terminating pods (have .metadata.deletionTimestamp) targeted by this deployment.
```
+1, updated

atiratree commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -633,11 +641,14 @@ func (dc *DeploymentController) syncDeployment(ctx context.Context, key string)
 	// List all Pods owned by this Deployment, grouped by their ReplicaSet.
 	// Current uses of the podMap are:
 	//
-	// * check if a Pod is labeled correctly with the pod-template-hash label.
```
This is an outdated comment and no longer applies. The podMap was used in other places in the past. Nowadays it is only used to check old pods running.

There was no change to `pod-template-hash` in this PR.

atiratree commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -633,11 +641,14 @@ func (dc *DeploymentController) syncDeployment(ctx context.Context, key string)
 	// Current uses of the podMap are:
 	//
-	// * check if a Pod is labeled correctly with the pod-template-hash label.
-	// * check that no old Pods are running in the middle of Recreate Deployments.
```
Once the feature is GA, we can depend on the rs.status.terminatingReplicas field, which is set even if the PodReplacementPolicy is unset. So we do not have to count the terminating replicas in the deployment controller anymore and can outsource the logic to the replicaset controller.

atiratree commented on cmd/kube-controller-manager/app/apps.go:
```diff
@@ -108,6 +108,7 @@ func startDeploymentController(ctx context.Context, controllerContext Controller
 		ctx,
 		controllerContext.InformerFactory.Apps().V1().Deployments(),
 		controllerContext.InformerFactory.Apps().V1().ReplicaSets(),
+		// TODO: remove the pod informer and pod permissions from deployment-controller's ClusterRole once DeploymentPodReplacementPolicy feature gate is graduated to GA and removed
```
please see: https://github.com/kubernetes/kubernetes/pull/123430#discussion_r1703976076 and https://github.com/kubernetes/kubernetes/pull/123430#discussion_r1703976720

atiratree commented on pkg/controller/deployment/recreate.go:
```diff
@@ -44,9 +47,20 @@ func (dc *DeploymentController) rolloutRecreate(ctx context.Context, d *apps.Dep
-
-	// Do not process a deployment when it has old pods running.
-	if oldPodsRunning(newRS, oldRSs, podMap) {
+	if utilfeature.DefaultFeatureGate.Enabled(features.DeploymentPodReplacementPolicy) {
```
Please see https://github.com/kubernetes/kubernetes/pull/123430#discussion_r1703976720. Terminating replicas are present unconditionally and we still wait for them when no policy is specified:

https://github.com/kubernetes/kubernetes/pull/123430/commits/a16b61914b3a5376fec085ae019c5678dca1b4a7#diff-c53d1f0a34c6462dc15af42d09038ed470c80ff7121fca5914ed2ccf60a8355aR56-R57

liggitt replied:
I'm pretty skeptical about modifying existing code paths for deployments which have not opted into this feature yet. There may be opportunity to do that in the future, but we've regressed complicated code paths in the past by significantly refactoring them for a new feature, and only discovered it after upgrading production clusters to the new code path.

atiratree replied:
If I look at this from a different perspective. I think this refactoring is quite useful as we can stop depending on pods and remove an informer. Now we can derive all the information we need from ReplicaSets. Also the `.status.terminatingReplicas` is available to everyone, even without opting into the feature. Suppose, we would postpone this refactoring to later, isn't it better to test it now, when it's feature gated?

atiratree replied:
I am not set on keeping the ungated logic and informer refactoring in. My points were mainly hypothetical. I agree with the downsides and we should pursue this later once we deem it reasonable.

Please see https://github.com/kubernetes/kubernetes/pull/123430#issuecomment-2455682729

atiratree commented on pkg/controller/deployment/recreate.go:
```diff
@@ -126,7 +141,16 @@ func oldPodsRunning(newRS *apps.ReplicaSet, oldRSs []*apps.ReplicaSet, podMap ma
-	scaled, _, err := dc.scaleReplicaSetAndRecordEvent(ctx, newRS, *(deployment.Spec.Replicas), deployment)
+func (dc *DeploymentController) scaleUpNewReplicaSetForRecreate(ctx context.Context, newRS *apps.ReplicaSet, deployment *apps.Deployment) (scaled bool, err error) {
+	newScale := *(deployment.Spec.Replicas)
+	if utilfeature.DefaultFeatureGate.Enabled(features.DeploymentPodReplacementPolicy) {
```
I actually had the policy check in this place before, but removed it to avoid code duplication, since the same check is done in the `NewRSNewReplicas` call. Also, the `NewRSNewReplicas` resolves to the same value (`deployment.Spec.Replicas`) if no policy is specified.

I agree, it is better to do this check twice to have clear code paths and better readability. I have updated this place to include the policy check.

atiratree commented on pkg/controller/deployment/util/deployment_util.go:
```diff
@@ -56,6 +58,11 @@ const (
 	// is deployment.spec.replicas + maxSurge. Used by the underlying replica sets to estimate their
```
The annotation is only used by the deployment, but it is set on the ReplicaSet object. I have updated the annotation description to make it more clear.

atiratree commented on pkg/controller/deployment/util/deployment_util.go:
```diff
@@ -56,6 +58,11 @@ const (
 	MaxReplicasAnnotation = "deployment.kubernetes.io/max-replicas"
+	// ReplicaSetReplicasBeforeScaleAnnotation is the number of replicas a replica set had before scaling began.
+	// The presence of this annotation indicates that a replica set has not yet fully scaled.
+	// The value is used by the underlying replica sets to estimate their scale proportions if the deployment has surge replicas
```
https://github.com/kubernetes/kubernetes/pull/123430#discussion_r1704145696

atiratree commented:
/remove-sig api-machinery

fedebongio commented:
/triage accepted

liggitt commented on pkg/controller/deployment/util/deployment_util.go:
```diff
@@ -56,6 +58,11 @@ const (
+	// The presence of this annotation indicates that a replica set has not yet fully scaled.
+	// The value is used by the underlying replica sets to estimate their scale proportions if the deployment has surge replicas
+	// (e.g. terminating pods) and the replica set is partially scaled due to the deployment's .spec.podReplacementPolicy.
+	ReplicaSetReplicasBeforeScaleAnnotation = "deployment.kubernetes.io/replicaset-replicas-before-scale"
```
all the relationships between them that are required for the controllers to work properly should be documented and tested

atiratree replied:
Okay, I can add that.

atiratree replied:
Thinking about it again, these are not relationships between the annotations. The annotations simply reflect the real world 
- `DesiredReplicasAnnotation`:  .spec.replicas of a deployment
- `MaxReplicasAnnotation`: .spec.replicas + maxSurge of a deployment
- `ReplicaSetReplicasBeforeScaleAnnotation`: .spec.replicas of a replica set

And the real world fields have relationships. For example .spec.replicas of a replica set is at most the .spec.replicas of a deployment. And we do have tests that the annotations correspond to the underlying fields.

Btw, I have updated the doc for `DesiredReplicasAnnotation` to specify what happens during scaling.

atiratree commented:
- rebased
- added more PodReplacementPolicy field and feature gate checks
- updated doc for DesiredReplicasAnnotation

liggitt commented:
I think there's three main threads intermixing in this PR:

1. Bookkeeping of the new status fields in ReplicaSet and Deployment API objects and controllers. Adding `status.terminatingReplicas` fields tracking terminating pods, and updating controller loops to track / clear that data (when the feature is enabled / disabled) would be an isolated change that could be implemented and reasoned about independently, and I agree is a prereq for anything else done for this feature.

2. Complicated controller code paths changing in ways that aren't easy to verify keep the current behavior when the feature gate is off or for deployments that have not opted into the new behavior. It is currently **very** hard to understand the overall controller change that happens when the feature is enabled. It is **very** important that the new deployment controller code paths around honoring the new pod replacement behaviors be isolated to when the new feature is enabled and the field is actually in use.
    * If I'm understanding the controller correctly, there's ~3 operations the controller does (rollout, scale up, scale down) and 2 replacement approaches the controller can take (TerminationStarted, TerminationComplete).
    * Deployments which don't specify an explicit podReplacementPolicy currently implicitly use one of the replacement approaches, depending on the operation and rollout strategy.
    * One way to make changes that are more clearly safe is to
        1. make the current code paths (dc.scale, dc.rolloutRecreate, dc.rolloutRolling) explicit about which behavior they implement (e.g. rename to `dc.scaleTerminationStarted`, `dc.rolloutRecreateTerminationComplete`, `dc.rolloutRollingTerminationStarted`), and leave the implementations of those methods untouched
        2. define three _new_ methods (e.g. `dc.scaleTerminationComplete`, `dc.rolloutRecreateTerminationStarted`, `dc.rolloutRollingTerminationComplete`)
        3. _only_ call those new methods when the feature is enabled _and_ the specific deployment has explicitly opted into the new behavior.

3. Use of Deployment-specific annotations on ReplicaSets as bookkeeping rather than teaching the replicaset controller about this TerminationStarted or TerminationComplete concept. I continue to think that it is questionable to try to implement this change only at the deployment controller level, but maybe with the new behaviors isolated in dedicated methods, it would be clearer that this is reasonable.

atiratree committed:
Implement PodReplacementPolicy for Deployments

Consider the TerminationStarted and TerminationComplete policies when
rolling out a new deployment with a Recreate or RollingUpdate strategy.

Also consider these policies when scaling any deployment, either during or
after a rollout has been completed. Special care has to be taken when scaling
up multiple replicasets proportionally. If a replicaset cannot be fully scaled,
"deployment.kubernetes.io/replicaset-replicas-before-scale" annotation is added
to the replicaset so that it can be scaled again in the future once the extra
pods terminate.

soltysh commented:
/hold
/lgtm cancel
/approve cancel 
based on Jordan's comment, let's try to divide this PR into smaller, more manageable chunks.

atiratree commented:
> Bookkeeping of the new status fields in ReplicaSet and Deployment API objects and controllers. Adding status.terminatingReplicas fields tracking terminating pods, and updating controller loops to track / clear that data (when the feature is enabled / disabled) would be an isolated change that could be implemented and reasoned about independently, and I agree is a prereq for anything else done for this feature.

I have split the feature into the https://github.com/kubernetes/kubernetes/pull/123430 and it can now be reviewed/merged separately.

> If I'm understanding the controller correctly, there's ~3 operations the controller does (rollout, scale up, scale down) and 2 replacement approaches the controller can take (TerminationStarted, TerminationComplete).
> make the current code paths (dc.scale, dc.rolloutRecreate, dc.rolloutRolling) explicit about which behavior they implement (e.g. rename to dc.scaleTerminationStarted, dc.rolloutRecreateTerminationComplete, dc.rolloutRollingTerminationStarted), and leave the implementations of those methods untouched

I have looked at changing all of these, but it would result in a lot of code duplication and it is not easy to split these neatly into such buckets (see the scaling logic and call graph).

**rolloutRecreate**:
- I have removed the pod informer refactoring.
- I have kept the original code flow and only using the new flow when non empty `spec.podReplacementPolicy` is specified.

**scale and rolloutRolling**:
- `rolloutRolling` is implemented via scaling, so the following points apply to both and we can focus on scaling.
- I have duplicated `scaleReplicaSet` to `scaleReplicaSetV2`, V2 then depends on the new annotation mutation functions `ComputeReplicaSetScaleAnnotationsV2` and `SetReplicaSetScaleAnnotationsV2`. This is guarded by the `DeploymentPodReplacementPolicy` feature gate. We rely on the new annotation functions to remove the `ReplicaSetReplicasBeforeScaleAnnotation` when it is not needed.
- I have also duplicated `scale` function into `scaleV2`. The V2 logic is quite similar to the original, but it has to account for transient pods and part of the logic is moved to `PrepareProportionalScalePlan`.

There are also a few places that are good to highlight:
- `NewRSNewReplicas`: The original code is left as is unless the feature gate is enabled and a `.spec.podReplacementPolicy` is specified. Nevertheless, there is some variable renaming.
- `getReplicaSetFraction`: The original code is left as is unless the feature gate is enabled and a `.spec.podReplacementPolicy` is specified.
- `SetNewReplicaSetAnnotations`: we use the annotation functions depending on the feature gate similar to scaling  

In summary, there should be no code path that is not gated by either the `DeploymentPodReplacementPolicy` feature and/or the presence of the `.spec.podReplacementPolicy` field. We switch between the V1/V2 logic to achieve better scaling to implement the feature (dependent on the `.spec.podReplacementPolicy`) and new annotation resolution (partially dependent on the `.spec.podReplacementPolicy`).

atiratree commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -354,6 +356,12 @@ func (dc *DeploymentController) deleteReplicaSet(logger klog.Logger, obj interfa
 // deletePod will enqueue a Recreate Deployment once all of its pods have stopped running.
 func (dc *DeploymentController) deletePod(logger klog.Logger, obj interface{}) {
+	// TODO: remove this handler, the pod informer and pod permissions from deployment-controller's ClusterRole once DeploymentPodReplacementPolicy feature gate is graduated to GA and removed
+	if utilfeature.DefaultFeatureGate.Enabled(features.DeploymentPodReplacementPolicy) {
```
I have removed the pod informer refactoring.

atiratree commented:
/test pull-kubernetes-e2e-gce

atiratree commented:
/test pull-kubernetes-e2e-gce

atiratree commented:
/test pull-kubernetes-e2e-gce

atiratree commented:
This PR now uses pointers for `.status.terminatingReplicas` as dicussed in https://github.com/kubernetes/kubernetes/pull/128546#discussion_r1832942828

atiratree committed:
TerminationComplete policy should wait for terminating pods to fully

terminate, before marking the deployment as complete

atiratree commented:
The first part introducing `.status.terminatingReplicas` has merged:  https://github.com/kubernetes/kubernetes/pull/128546

atiratree commented:
@liggitt can we please continue with this PR review?

atiratree commented:
/retest

atiratree commented:
The PR has been rebased and is now gated by two feature gates DeploymentReplicaSetTerminatingReplicas and DeploymentPodReplacementPolicy.

elieser1101 commented:
Hello @liggitt @atiratree
Appreciate all of your efforts with this PR! Seem like the plan  is still to resolve it for v1.34 release
This is a gentle reminder that the code freeze will start at  [02:00 UTC Friday 25th July 2025](https://everytimezone.com/s/a2c01c54) . Please make sure any PRs have both lgtm and approved labels ASAP, and file an [Exception](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md) if you think this PR needs additional time.
Thanks!

soltysh commented on pkg/features/kube_features.go:
```diff
@@ -1176,6 +1183,10 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
+
 	DeploymentReplicaSetTerminatingReplicas: {
 		{Version: version.MustParse("1.33"), Default: false, PreRelease: featuregate.Alpha},
 	},
```
Further below there's `defaultKubernetesFeatureGateDependencies` where we express dependencies between FGs. Since your new feature relies on `DeploymentReplicaSetTerminatingReplicas` it would be good to define it there.

atiratree replied:
That is very cool that we can declare dependencies now. Updated.

soltysh commented on pkg/registry/apps/deployment/strategy_test.go:
```diff
@@ -41,6 +42,213 @@ const (
+
+	withTwoFeatureGates := func(testName string, enableDeploymentPodReplacementPolicy bool, test func(testName string, deploymentReplicaSetTerminatingReplicasEnabled, deploymentPodReplacementPolicyEnabled bool)) {
+		for _, deploymentReplicaSetTerminatingReplicasEnabled := range []bool{true, false} {
+			for _, deploymentPodReplacementPolicyEnabled := range []bool{true, false} {
```
I wonder, now that we have the ability to ensure FG dependencies, do we need to tests all variants of both FG on/off?

soltysh commented on pkg/controller/deployment/sync.go:
```diff
@@ -400,15 +409,185 @@ func (dc *DeploymentController) scale(ctx context.Context, deployment *apps.Depl
+	}
 }
 
+func (dc *DeploymentController) scaleReplicaSetV2(ctx context.Context, rs *apps.ReplicaSet, newScale int32, deployment *apps.Deployment, partialScaling bool) (bool, *apps.ReplicaSet, error) {
```
This `scaleReplicaSetV2` is worrisome. This will kick in every time we have the FG enabled, so we have two possible paths for the same functionality (the case when FG is on and no podReplacementPolicy is set, or when the FG is off). The only difference I see is two function invocations: 
1. `ComputeReplicaSetScaleAnnotationsV2` vs `ReplicasAnnotationsNeedUpdate`
2. `SetReplicaSetScaleAnnotationsV2` vs `SetReplicasAnnotations`

I think it'd be easier to digest if we modify the current path such that it looks more like the new proposed functions. Then the additional changes would be gated through the new FG, only.

atiratree replied:
Personally, I would prefer to unify the annotation utility functions as well, but this would involve touching the existing code paths. Thus increasing a chance of a regression.

I would like to hear @liggitt opinion on the annotations before proceeding any further.

soltysh reviewed (commented):
I've only carefully reviewed the API parts (that part lgtm as an API shadow). 

But for the controller changes I've only went through the recreate paths, I'm going to be reviewing the remaining paths in the upcoming days. But already I see a few items that need to handled before we even attempt this:
1. Unit tests for `rolloutRecreate` is a must, if we can't proof the current functionality is working correctly, I'm not comfortable changing this path.
2. The `*V2` functions, specifically `scaleReplicaSetV2` (like I said earlier I haven't get to them all) is a no-go, the fact that we're introducing two paths for the same code flow is suspicious. I'd rather see the current functions be modified towards what you're doing in the newer versions, such that the gating will only affect minor pieces of the code.

soltysh commented:
/priority important-longterm

atiratree commented on pkg/controller/deployment/recreate.go:
```diff
@@ -44,9 +46,26 @@ func (dc *DeploymentController) rolloutRecreate(ctx context.Context, d *apps.Dep
+		// Do not proceed with a deployment when it has old pods running.
+		if oldPods := util.GetActualReplicaCountForReplicaSets(oldRSs); oldPods > 0 {
+			return dc.syncRolloutStatus(ctx, allRSs, newRS, d)
+		}
```
The terminating replicas should produce the same results as the `oldPodsRunning`, but with the added benefit of not needing the pod lister and building the pod map. To be safe, terminating replicas field is used only when the `podReplacementPolicy` feature is used. This enables us to gain confidence and switch to using terminating replicas later and ultimately remove the lister/pod usage.

atiratree commented:
> The *V2 functions, specifically scaleReplicaSetV2 (like I said earlier I haven't get to them all) is a no-go, the fact that we're introducing two paths for the same code flow is suspicious. I'd rather see the current functions be modified towards what you're doing in the newer versions, such that the gating will only affect minor pieces of the code.

Please note the above discussion regarding this topic 
- https://github.com/kubernetes/kubernetes/pull/123430#issuecomment-2452537034
- https://github.com/kubernetes/kubernetes/pull/123430#issuecomment-2455682729

Originally, I included the gated code in the places where it was necessary within the original code paths, but @liggitt had objected to touching the existing code paths to make sure the current behavior is not broken by this feature.

Note: a similar example of this can be found in StatefulSets: https://github.com/kubernetes/kubernetes/blob/370c2f1d9846231fdb2d51c83ab0932df5fa0642/pkg/controller/statefulset/stateful_set_control.go#L658-L659
